### ISSUE TITLE
Mounting several external storages over the same mount point causes the "old" storages to get overwritten

### ISSUE THREAD

**Issue by jvillafanez** [owncloud/core#10511]
### Steps to reproduce
1. Mount a external storage in folder "F1"
2. Mount another external storage in the same folder "F1"
3. The first external storage won't be accessible because the "F1" folder will access to the last external storage

Note: the same happens if a local folder "F1" is already created and an external storage is mounted there
### Expected behaviour

Either core throws an error telling that it's not possible to mount the external storage in that location (so the app can handle this error by its own) or core rename the mount automatically
### Actual behaviour

The last external storage overwrites the other storages making them inaccessible

@MTRichards

**Comment by SergioBertolinSG** [comment]
This is still happening with external storages. Also with windows network drive.
I think mounted folder should be automatically renamed if there is one already mounted with that name or throw an error when mounting warning the user for choosing a different name.

**Comment by PVince81** [comment]
Yes, and as long as all mount points aren't specified through the external storage app (which means WND/Sharepoint app ported to files_external), it will be difficult if not impossible to implement this kind of logic. Well, or we need to implement it in 3-4 different places at once, which is not desirable.

**Comment by PVince81** [comment]
Moving to 8.2

**Comment by PVince81** [comment]
Depends on https://github.com/owncloud/core/issues/16304

**Comment by SergioBertolinSG** [comment]
OK, 8.2 then.

**Comment by PVince81** [comment]
@Xenopathic any new insights on this ? Would the priority mechanism help with this somehow ?

**Comment by PVince81** [comment]
Whenever sharepoint/wnd get integrated with files_external, they will inherit the priority behavior. This means that overwriting a mount point will still be possible, but a priority logic can be applied.

Not sure which behavior is the most desirable.

**Comment by Xenopathic** [comment]
The priority logic really only applies in the admin settings. Before, the storage that took precedence was basically arbitrary, but now it follows: user > group > all. So in that regard, it works a lot better. However, there is still an issue that if you create a storage mount applicable to a few users (e.g. 'userA', 'userB', 'userC'), then create a second mount with the same mountpoint and with one of the same users (e.g. 'userB', 'userD'), then the first storage will have the repeated user removed from it. This is far from optimal, but it's a limitation of our current mount.json format.

In personal settings, you can only create a single mount with a given mountpoint. Trying to create a second one will refuse to save it.

**Comment by DeepDiver1975** [comment]
@karlitschek @cmonteroluque to fix this issue we require #16304 which is scheduled for 9.0 - https://github.com/owncloud/core/issues/10511#issuecomment-112476155

We should move this issue as well.

**Comment by schiesbn** [comment]
I just skim through the issue. But shouldn't be a solution relatively easy to implement similar to what we do in sharing: https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/sharedmount.php#L70

Basically on mount we do a ````files_exists($mountPoint)````, if it returns false we mount it, if it returns true we generate a new unique target, write it to the share table (in this case we would update the external storage config) and mount it with the new name.

**Comment by PVince81** [comment]
The problem with renaming mount points on-demand is that concurrency might execute the same logic several times at once. But back then we didn't have locking, maybe locking the local mount point first could work.

**Comment by PVince81** [comment]
@schiesbn feel free to take over if you'd like to have a try with this approach

**Comment by PVince81** [comment]
Dynamically renaming mount points (adding a "(2)") at mount time will not work with the WND/Sharepoint apps which have their own config formats.

**Comment by PVince81** [comment]
The current configuration format makes it tricky.

Also the question is whether to catch this at configuration time (best) or once it is already mounted.

It all depends on the type of mount points.

#### System-wide mount points
If we're talking about the admin setting up mount points, then
1) if the mount point has the same name and the configuration is 100% identical except with different "applicable" field, then the applicable fields are merged (this is due to how the config works internally)

2) if the mount point has the same name but the configuration is **not** 100% identical, it currently seems to happily override the mount point. I'm guessing that this is what this ticket is about.

#### Personal mount points
1) Same mount point name and same config should probably get merged automatically (not tested) but also not a bad case

2) Same mount point name and different config, the mount point gets overridden by the last config. I guess this is also what this ticket is about

#### System mount exist, personal mount overrides
Here is the question whether a user is allowed to override a system mount with an own mount point

#### Personal mount exists, admin mounts a new storage using the same name
This is https://github.com/owncloud/core/issues/11794

So basically we have three completely different cases:
- [ ] a second system-wide mount with different config should not be allowed to override an existing system-wide mount
- [ ] a second personal-wide mount point with different config should not be allowed to override an existing personal mount
- [ ] a personal mount should not be allowed to override a system-wide mount (or even an existing folder ?)
- [ ] a newly created system-wide mount should not be allowed to override an existing person mount (or even an existing folder ?), this is very complicated and tracked here https://github.com/owncloud/core/issues/11794

So which one of these is "high" and the most important / annoying ?

**Comment by PVince81** [comment]
The solution proposed by @schiesbn could cover all the cases at once. It means that the UI will not show any error but the mount point name will be automatically changed to "(2)" at mount time. I tried this before in this PR https://github.com/owncloud/core/pull/11807 but is causing trouble and potential performance issues and was closed after an advice from @icewind1991 

Needs further research.

**Comment by cmonteroluque** [comment]
OK, 9.0. We're moving a lot of external storage work into it, we will have to plan this carefully

**Comment by PVince81** [comment]
@icewind1991 I suppose the new "oc_mounts" table might help detect collisions better ?

**Comment by PVince81** [comment]
Still happens on master fae6643e6dd4bc308170e0ab45886f110f60a639.

The GUI doesn't check for duplicate mount point names so I was able to mount two system-wide SFTP mounts onto the same mount point "/sftp".

**Comment by PVince81** [comment]
In the GUI it is also tricky to detect duplicates, because it could be a legitimate use case, for example:
- for users user1 and user2, mount SFTP with config1 as "/sftp"
- for users user3 and user4, mount SFTP with config2 as "/sftp"

Such configs don't overlap. To detect overlaps, the JS code would have to check whether there is an overlap in user names from the applicable field.

Now if we want to detect this on the server side, it means checking the "oc_mounts" table of all users to see if any of them already has a mount point at this location. Trouble is: oc_mounts only updates itself for that user after the user's FS has setup (after accessing files / logging in, etc).
Even that might not be enough because they might already have a folder there.

**Comment by PVince81** [comment]
Note that now we also have the CLI command to create mount points, so whichever solution is implemented would need to cover both the GUI configs and CLI config.

**Comment by PVince81** [comment]
Let's see, maybe there is a way to do this detection on database config level: before saving, check whether there is already another mount point with the same name and intersecting applicables.

**Comment by PVince81** [comment]
Hmmm, this could happen too:
- for users user1 and user2, mount SFTP with config1 as "/sftp"
- for users user3 and user4, mount SFTP with config2 as "/sftp"

Then the admin tries adding "user2" to the second config. This would create an overlap, in which case we can't let the admin save it. The admin would have to create a third config for "user2" and set the mount point manually to something else than "/sftp".

Tricky tricky.

**Comment by PVince81** [comment]
What if we just allowed overlaps but converted the visible mount point name of the user to a renamed name like "/sftp (2)" ?
That would probably be the best approach. However there is no good place in the database to save the user's own name for the mount point.

@icewind1991 how about adding a column "mount_id" to "oc_mounts" to map to the ext storage configuration ? Then whenever the user's FS is being setup, we can check for duplicates and write the actual mount point name into the "oc_mounts" table and keep it there.

This would also bring the advantage of letting users rename external storage mount points for themselves (just like received shares).

**Comment by cmonteroluque** [comment]
@PVince81 @MTRichards moving this to 9.1 as we look into files_external then?

**Comment by MTRichards** [comment]
Yes. Seems to solve a problem, and may be wise to work it into files external since we are consolidating there for consistent handling of issues.

**Comment by PVince81** [comment]
Yeah. I hope we can make some more changes/additions under the hood to make such check possible (oc_mounts table)

**Comment by PVince81** [comment]
Moving to backlog.

The potential solution is overly complex due to not fitting architecture and this issue doesn't seem to be a major blocker for any one so far.

Ideally we should rework the FS mount system as described in https://github.com/owncloud/core/issues/26190 which would make it possible to catch conflicts in a central place.

**Comment by ownclouders** [comment]
Hey, this issue has been closed because the label `status/STALE` is set and there were no updates for 7 days. Feel free to reopen this issue if you deem it appropriate.

(This is an automated comment from [GitMate.io](https://gitmate.io).)

**Comment by PVince81** [comment]
Leaving closed. Requires proper mount manager. See https://github.com/owncloud/core/issues/26190
